Use GIL to guard py::object's destruction

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38376 Use GIL to guard decref of jit::toPyObj return value in processRpc
* **#38348 Use GIL to guard py::object's destruction**
* #38366 Explicitly decref py::object in PythonRpcHandler
* #38364 Explicitly decref py::object in ConcretePyObjectHolder and PythonFunctionGuard
* #38340 Minor code cleanup



Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

Commit:
Keep py::object alive until jit::toIValue returns

[ghstack-poisoned]

Commit:
Update on "Keep py::object alive until jit::toIValue returns"




[ghstack-poisoned]

mrshenli commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -476,6 +476,17 @@ inline void guardAgainstNamedTensor(const T& var) {
+// keep the py::objec alive until this function returns.
+//
+// Note [jit::toIValue barrow py::object refcnt]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Hey @ezyang, I am not sure if my understanding is correct here. Is it true that if we pass a `py::object` to `toIValue`, we need to make sure that this `py::object` is alive until `toIValue` returns, because py::handle here does not inc the refcnt?

wanchaol replied:
Yes this is correct, because `toIValue` mostly used as function argument conversion, and it is almost certain that the caller manages the lifetime of the `py::handle` passed in.

rohan-varma replied:
Might be a naive question, but what's the problem with changing `toIValue` to call `inc_ref` to keep the underlying object alive for the duration of the function call, or even just use `py::object` here? This would make it a lot easier for callers and potentially avoid bugs/flakiness. Are we concerned about the performance implications?

jjlilley replied:
Don't we care that the value of the `py::object` is persisted for the duration of `markComplete()`, not just `jit::toIValue()`?

mrshenli replied:
If below is the code that runs when passing a py::object to py::handle, it actually inc the refcnt?

https://github.com/pybind/pybind11/blob/a86ac538f59db7af596ee192f42b9f39ec8039ed/include/pybind11/cast.h#L1623-L1625

```c++
struct pyobject_caster {
    template <typename T = type, enable_if_t<std::is_same<T, handle>::value, int> = 0>
    bool load(handle src, bool /* convert */) { value = src; return static_cast<bool>(value); }

    template <typename T = type, enable_if_t<std::is_base_of<object, T>::value, int> = 0>
    bool load(handle src, bool /* convert */) {
        if (!isinstance<type>(src))
            return false;
        value = reinterpret_borrow<type>(src);
        return true;
    }

    static handle cast(const handle &src, return_value_policy /* policy */, handle /* parent */) {
        return src.inc_ref();
    }
    PYBIND11_TYPE_CASTER(type, handle_type_name<type>::name);
};
```

mrshenli replied:
> Don't we care that the value of the py::object is persisted for the duration of markComplete(), not just jit::toIValue()?

`jit::toIValue` should have created a copy of the `py::object` in the cast below. So as long as the IValue is alive, the py::object would also be alive I think?

https://github.com/pytorch/pytorch/blob/ee1ddcef8d2994b59c9d7aeae8f0dd455ee6e469/torch/csrc/jit/python/python_ivalue.h#L17-L21

mrshenli replied:
> Might be a naive question, but what's the problem with changing toIValue to call inc_ref to keep the underlying object alive for the duration of the function call, or even just use py::object here? This would make it a lot easier for callers and potentially avoid bugs/flakiness. Are we concerned about the performance implications?

Not sure what are the reason for using py::handle here. @wanchaol is there any specific reason for this?

wanchaol replied:
I don't remember if there's a specific reason to keep a `py::handle`, merely to avoid redundant inc_refs as arguments are suppose to be alive when you pass to a function.

cc @zdevito who might have more context on this.

xush6528 replied:
Say we have
```cpp
py::object o;

py::handle aa(o);  // copy constructor
py::handle aaa = o;  // copy constructor, as it's also initialization.
```
The constructor uses here is the synthesized default copy-constructor, `py::handle(const py::handle&)`.
`py::object` is statically casted to `const py::handle&` type as it's a subtype.
Also the lifetime of the RValue(temporary) `py::object` returned by `toPyPObject(..)` is prolonged by binding to the const reference (`const py::handle&`). 

Similar for,
```cpp
py::handle bb; // The empty constructor of py::handle is called.
bb = o;  // copy assign operator
```
The synthesized default copy-assign-operator, `py::handle::operator=(const py::handle&)`.

Commit:
Update on "Keep py::object alive until jit::toIValue returns"


Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]

Commit:
Update on "Keep py::object alive until jit::toIValue returns"


Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -109,7 +109,10 @@ PyRRef::PyRRef(const py::object& value, const py::object& type_hint)
         TypePtr elem_type = tryInferTypeWithTypeHint(value, type_hint);
         auto rref = RRefContext::getInstance().createOwnerRRef(elem_type);
         py::object copy(value); // increases refcount
```
Hey @wanchaol Do you know when we do `std::move` in the original code below, what happens to the reference cnt on the `py::object`? 

And why do we need the copy here? The `jit::toIValue` below should have already inc the refcnt, no?

ezyang replied:
When you `move` copy into `toIValue`, `toIValue` *may* take ownership of copy, and you would no longer have a lifetime guarantee through value. However, because you are calling `PyRRef` with a const reference to `value`, the general assumption is that this means that `value` will stay live for the entirety of the function.

xush6528 replied:
Compiler only synthesizes move constructor for a class that has all its non-static members movable (have move constructor). 
Since `py::handle` has a non-movable member, `PyObject *m_ptr` (https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L217), `py::handle` does not have a synthesized move constructor. 

When a class does not have a move constructor defined, but an RValue is passed to its constructor, the copy constructor is called in stead.
So these two calls are equivalent. The case converges to the one mentioned above.
```cpp
IValue ivalue = jit::toIValue(copy, elem_type);
IValue ivalue = jit::toIValue(std::move(copy), elem_type);
```

rohan-varma commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -476,6 +476,17 @@ inline void guardAgainstNamedTensor(const T& var) {
+// passes a py:object as the first argument, this will be copy by value and the
+// py::handle is only live as long as py::object is, and the py::handle is not
+// reference counted. Hence, the call site needs to keep py::object alive until
+// a new copy is created by jit::toIValue().
```
Could this just say "keep py::object alive until toIValue()" returns? The caller should not have to be responsible for knowing the internal details, right?

mrshenli replied:
Yes, let me modify that.

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -109,7 +109,10 @@ PyRRef::PyRRef(const py::object& value, const py::object& type_hint)
         py::object copy(value); // increases refcount
-        IValue ivalue = jit::toIValue(std::move(copy), elem_type);
+        // Keep copy alive until jit::toIValue returns.
+        // See Note [jit::toIValue barrow py::object refcnt] at jit::toIValue
```
barrow -> borrow

jjlilley commented:
Wondering if this might fix a "Fatal Python error: GC object already tracked" (P130801732) error that I'm seeing on the master with prod-preset feed run?

(I never saw it before Friday, but saw it happen top of tree right after Shihao's change on Fri night, though could be unrelated)

ezyang commented:
I had a brief 1:1 chat with @mrshenli and I wanted to remind everyone of some important things to remember when doing multithreading and Python GIL stuff:

1. Do not put smart Python C++ wrapped objects in structs that do not require GIL protection. They will get destructed when those structs get freed, EVEN if you explicitly write a destructor (destructor does not replace the element-wise destruction of each field!) It is much safer to store a raw pointer.
1a. Do not put smart Python C++ wrapped objects in a scope where you don't also have a GIL taken out. The object will get destructed at the end of the scope, and when that destruction happens you need to have the GIL.
2. You must take out the GIL whenever you read/write Python objects. In single-threaded settings this is hard to get wrong, but in multithreading you have to be very careful.

Commit:
Update on "Keep py::object alive until jit::toIValue returns"


Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]

Commit:
Update on "Keep py::object alive until jit::toIValue returns"


Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]

Commit:
Update on "Use GIL to guard py::object's destruction"




Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -24,6 +24,7 @@ namespace rpc {
 namespace {
 
 py::object toPyObj(const Message& message) {
+  py::gil_scoped_release release;
```
For which call this is?

mrshenli replied:
Yes, it was for PythonRpcHandler. I updated the PR to remove `toPyObj` and added a `toIValue`, as I think former is no longer needed.

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -32,7 +33,7 @@ py::object toPyObj(const Message& message) {
       stack.push_back(ret.value());
       {
-        pybind11::gil_scoped_acquire ag;
+        pybind11::gil_scoped_acquire acquire;
```
"pybind11" has alias "py". 
https://github.com/pytorch/pytorch/blob/11a40410e755b1fe74efe9eaa635e7ba5712846b/torch/csrc/utils/pybind.h#L18

We can make it shorter. `py::gil_scoped_acquire acquire;`

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -118,14 +119,18 @@ c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
-                PyObjectType::get()));
+            IValue value;
+            {
+              // Need GIL to destruct the py::object returned by toPyObj
```
I see.

Before https://github.com/pytorch/pytorch/pull/37519,
the code was like this, https://github.com/pytorch/pytorch/blame/580928801f5d4c229cb33b117f152af46e7a2a82/torch/csrc/distributed/rpc/init.cpp#L340
The `py::object` returned by `toPyObj(..)` was passed to Python directly. pybind11 takes care of its destruction.

While after the PR, this `toPyObj(..)` returned `py::object` is passed to `toIValue(..)` as `const py::handle&`, from which `PyObjectHolder` creates a new `py::object` copy. https://github.com/pytorch/pytorch/blob/dc28784f50da8276d87bfb7ed0b990aa88bc4ff0/torch/csrc/jit/python/python_ivalue.h#L20

With the destruction of the new copy taken care of by https://github.com/pytorch/pytorch/blob/dc28784f50da8276d87bfb7ed0b990aa88bc4ff0/torch/csrc/jit/python/python_ivalue.h#L29.

Now we have to take care of destruction of the old copy.

mrshenli replied:
Yes, exactly! 

There might be other places that need some change, e.g., UnpickledPythonCall holds a py::object, but relies on caller to call the `movePythonUdf`. Will fix this on top of this PR.

Commit:
Update on "Use GIL to guard py::object's destruction"




Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -32,17 +32,25 @@ py::object toPyObj(const Message& message) {
       Stack stack;
       stack.push_back(ret.value());
       {
```
We don't need this bracket?

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -32,17 +32,25 @@ py::object toPyObj(const Message& message) {
       auto& resp = static_cast<PythonResp&>(*response);
       auto& pythonRpcHandler = PythonRpcHandler::getInstance();
-      return pythonRpcHandler.deserialize(resp.serializedPyObj());
+      {
```
Same

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -32,17 +32,25 @@ py::object toPyObj(const Message& message) {
+        // Need GIL to destruct the py::object returned by deserialize()
+        py::gil_scoped_acquire acquire;
+        return jit::toIValue(
+            pythonRpcHandler.deserialize(resp.serializedPyObj()),
```
It works because `py::gil_scoped_acquire` is constructed before the temp `py::object` is returned by `pythonRpcHandler.deserialize(..)`, they are destructed in reverse order.

Commit:
Update on "Use GIL to guard py::object's destruction"




Differential Revision: [D21530282](https://our.internmc.facebook.com/intern/diff/D21530282)

[ghstack-poisoned]
